File: v2/src/planner.ts
/**
 * Planner — turns a single high-level goal into a DAG of OpenHands tasks.
 *
 * Cognition runs in Claude (we shell out to `claude --print`); the TS layer
 * only validates the JSON the model produces. Same split as the OpenHands
 * runner: mechanism here, decisions in Claude.
 *
 * The planner returns a topologically-ordered list of nodes. The caller
 * (TaskQueue.submitProgram) assigns UUIDs and remaps the model-supplied
 * keys to real task ids before persisting.
 *
 * Claude CLI shape:
 *   claude --print --output-format=text "<prompt>"
 * Auth comes from the bind-mounted ~/.claude session — no API key, ever.
 *
 * Single goals (the "small bounded" case from architecture-v2 §3.2) still
 * go through here: the planner returns a 1-element list with no deps. One
 * code path; less special-casing in the queue.
 */

import { spawn } from "node:child_process";

import type { TaskKind } from "./state-store.js";

const PLANNER_TIMEOUT_MS = Number(process.env["DEVCLAW_PLANNER_TIMEOUT_MS"] ?? "90000");
const CLAUDE_BIN = process.env["DEVCLAW_CLAUDE_BIN"] ?? "claude";
const MAX_TASKS_PER_PLAN = 20;

const VALID_KINDS: TaskKind[] = ["implement_feature", "fix_bug", "review_repository"];

export type PlannedTask = {
  /** Stable model-assigned id used to express deps within this plan only. */
  key: string;
  goal: string;
  kind: TaskKind;
  /** Keys (not UUIDs) of other tasks in this plan that must finish first. */
  dependsOnKeys: string[];
};

export class PlannerError extends Error {
  constructor(message: string, public readonly raw?: string) {
    super(message);
    this.name = "PlannerError";
  }
}

const SYSTEM_PROMPT = `You are DevClaw's planner. Decompose a single coding goal
into a directed acyclic graph (DAG) of smaller tasks that can each be executed
by an autonomous coding agent in one run.

Rules:
- Each task is bounded: an agent should finish it in one session.
- Prefer fewer, larger tasks over many tiny ones. Aim for 1–6 tasks. Use more
  only when the goal is genuinely large.
- If the goal is small (e.g. "fix a typo", "add a config flag"), return ONE task.
- Use "depends_on" for tasks that genuinely cannot start until another finishes
  (e.g. "frontend uses the API contract from task 1"). Don't invent fake deps.
- Independent tasks should have empty depends_on so they can run in parallel.
- Task "kind" must be one of: implement_feature, fix_bug, review_repository.
  Default to implement_feature unless the goal explicitly says fix a bug or
  review code without changing it.

Respond with STRICT JSON ONLY — no prose, no markdown fences. Schema:

{
  "tasks": [
    {
      "key": "<short stable id, e.g. 't1', 'scaffold'>",
      "goal": "<concrete instruction for the agent>",
      "kind": "implement_feature" | "fix_bug" | "review_repository",
      "depends_on": ["<key of another task in this plan>", ...]
    }
  ]
}`;

export function buildPlannerPrompt(goal: string, workspaceDir: string): string {
  return `${SYSTEM_PROMPT}

Workspace: ${workspaceDir}
Goal: ${goal}

Return the JSON now.`;
}

/**
 * Pull the first JSON object out of a model response. Tolerates leading
 * prose or markdown fences even though the prompt forbids them.
 */
export function extractJson(text: string): string {
  const trimmed = text.trim();
  if (trimmed.startsWith("{")) return trimmed;
  const fence = trimmed.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
  if (fence && fence[1]) return fence[1];
  const first = trimmed.indexOf("{");
  const last = trimmed.lastIndexOf("}");
  if (first >= 0 && last > first) return trimmed.slice(first, last + 1);
  throw new PlannerError("No JSON object found in planner response", text);
}

/**
 * Validate the parsed plan and return tasks in topological order.
 * Throws PlannerError on cycles, dangling refs, missing fields, etc.
 */
export function validatePlan(parsed: unknown): PlannedTask[] {
  if (!parsed || typeof parsed !== "object") {
    throw new PlannerError("Plan must be a JSON object");
  }
  const obj = parsed as Record<string, unknown>;
  const raw = obj["tasks"];
  if (!Array.isArray(raw)) {
    throw new PlannerError("Plan.tasks must be an array");
  }
  if (raw.length === 0) {
    throw new PlannerError("Plan must contain at least one task");
  }
  if (raw.length > MAX_TASKS_PER_PLAN) {
    throw new PlannerError(
      `Plan has ${raw.length} tasks; max is ${MAX_TASKS_PER_PLAN}. Refine the goal.`,
    );
  }

  const seen = new Set<string>();
  const tasks: PlannedTask[] = [];
  for (const t of raw) {
    if (!t || typeof t !== "object") {
      throw new PlannerError("Each task must be an object");
    }
    const r = t as Record<string, unknown>;
    const key = typeof r["key"] === "string" ? (r["key"] as string).trim() : "";
    const goal = typeof r["goal"] === "string" ? (r["goal"] as string).trim() : "";
    const kindRaw = typeof r["kind"] === "string" ? (r["kind"] as string) : "implement_feature";
    const depsRaw = r["depends_on"];
    if (!key) throw new PlannerError("Task missing 'key'");
    if (!goal) throw new PlannerError(`Task '${key}' missing 'goal'`);
    if (seen.has(key)) throw new PlannerError(`Duplicate task key '${key}'`);
    if (!VALID_KINDS.includes(kindRaw as TaskKind)) {
      throw new PlannerError(
        `Task '${key}' has invalid kind '${kindRaw}'; expected one of ${VALID_KINDS.join(", ")}`,
      );
    }
    const dependsOnKeys: string[] = [];
    if (depsRaw != null) {
      if (!Array.isArray(depsRaw)) {
        throw new PlannerError(`Task '${key}' depends_on must be an array`);
      }
      for (const d of depsRaw) {
        if (typeof d !== "string" || !d.trim()) {
          throw new PlannerError(`Task '${key}' has non-string dep`);
        }
        if (d === key) {
          throw new PlannerError(`Task '${key}' depends on itself`);
        }
        dependsOnKeys.push(d.trim());
      }
    }
    seen.add(key);
    tasks.push({ key, goal, kind: kindRaw as TaskKind, dependsOnKeys });
  }

  // Validate all dep refs resolve.
  for (const t of tasks) {
    for (const d of t.dependsOnKeys) {
      if (!seen.has(d)) {
        throw new PlannerError(`Task '${t.key}' depends on unknown key '${d}'`);
      }
    }
  }

  // Kahn topological sort — also detects cycles.
  const byKey = new Map(tasks.map((t) => [t.key, t]));
  const indegree = new Map<string, number>();
  const dependents = new Map<string, string[]>();
  for (const t of tasks) {
    indegree.set(t.key, t.dependsOnKeys.length);
    for (const d of t.dependsOnKeys) {
      const list = dependents.get(d) ?? [];
      list.push(t.key);
      dependents.set(d, list);
    }
  }
  const ready: string[] = [];
  for (const [k, n] of indegree) if (n === 0) ready.push(k);
  ready.sort(); // deterministic order across runs

  const ordered: PlannedTask[] = [];
  while (ready.length) {
    const k = ready.shift() as string;
    ordered.push(byKey.get(k) as PlannedTask);
    const downstream = dependents.get(k) ?? [];
    for (const d of downstream) {
      const n = (indegree.get(d) ?? 0) - 1;
      indegree.set(d, n);
      if (n === 0) ready.push(d);
    }
    ready.sort();
  }
  if (ordered.length !== tasks.length) {
    throw new PlannerError("Plan contains a dependency cycle");
  }
  return ordered;
}

/**
 * Spawn `claude --print` with the planner prompt and return its stdout.
 * Exported so tests can stub it.
 */
export async function callClaude(prompt: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    const env: NodeJS.ProcessEnv = { ...process.env };
    // Belt + suspenders: never let an API key on the process env override
    // the OAuth session in ~/.claude.
    delete env.ANTHROPIC_API_KEY;
    delete env.ANTHROPIC_AUTH_TOKEN;

    const child = spawn(
      CLAUDE_BIN,
      ["--print", "--output-format=text", prompt],
      { env, stdio: ["ignore", "pipe", "pipe"] },
    );

    let stdout = "";
    let stderr = "";
    let timedOut = false;
    const timer = setTimeout(() => {
      timedOut = true;
      child.kill("SIGKILL");
    }, PLANNER_TIMEOUT_MS);

    child.stdout.on("data", (b: Buffer) => {
      stdout += b.toString();
    });
    child.stderr.on("data", (b: Buffer) => {
      stderr += b.toString();
    });
    child.on("error", (err) => {
      clearTimeout(timer);
      reject(new PlannerError(`Failed to spawn ${CLAUDE_BIN}: ${err.message}`));
    });
    child.on("close", (code) => {
      clearTimeout(timer);
      if (timedOut) {
        reject(
          new PlannerError(
            `claude --print timed out after ${PLANNER_TIMEOUT_MS}ms. stderr:\n${stderr}`,
          ),
        );
        return;
      }
      if (code !== 0) {
        reject(
          new PlannerError(
            `claude --print exited ${code}. stderr:\n${stderr}`,
            stdout,
          ),
        );
        return;
      }
      resolve(stdout);
    });
  });
}

/**
 * Full planner entry point: prompt Claude, validate, return ordered DAG.
 * `claudeCaller` is injected so tests can stub the subprocess.
 */
export async function planGoal(
  goal: string,
  workspaceDir: string,
  claudeCaller: (prompt: string) => Promise<string> = callClaude,
): Promise<PlannedTask[]> {
  const prompt = buildPlannerPrompt(goal, workspaceDir);
  const raw = await claudeCaller(prompt);
  const jsonText = extractJson(raw);
  let parsed: unknown;
  try {
    parsed = JSON.parse(jsonText);
  } catch (err) {
    throw new PlannerError(
      `Planner JSON parse failed: ${(err as Error).message}`,
      raw,
    );
  }
  return validatePlan(parsed);
}

File: devclaw/planner.py
"""Planner — turns a single high-level goal into a DAG of OpenHands tasks.

Cognition runs in Claude (we shell out to ``claude --print``); this layer only
validates the JSON the model produces. Same split as the runner: mechanism
here, decisions in Claude. Auth comes from the bind-mounted ~/.claude session —
no API key, ever.

Single goals (the "small bounded" case) still go through here: the planner
returns a one-element list with no deps. One code path; less special-casing.
"""

from __future__ import annotations

import asyncio
import json
import sys
from dataclasses import dataclass, field
from typing import Awaitable, Callable

from .state_store import TaskKind

# The low-level LLM-call primitive lives in llm_call.py — a LEAF module
# (extracted 2026-07-19) so the quality gate imports the call primitive
# without dragging this module's state_store/task_git deps behind it.
# Re-exported verbatim: every historical `from .planner import …` call site —
# and tests patching `planner.call_claude` (cognition.py resolves it lazily
# through THIS namespace) — keeps working unchanged.
from .llm_call import (  # noqa: F401
    _COGNITION_TIMEOUT_DEFAULT_S,
    _build_claude_argv,
    _cognition_timeout_ms_from_env,
    CLAUDE_BIN,
    CliEnvelope,
    PLANNER_TIMEOUT_MS,
    PlannerError,
    call_claude,
    claude_with_model,
    extract_json,
    parse_cli_envelope,
)

# Imported as a module global (not called via task_git.) so tests can patch it
# on THIS namespace — same convention as task_queue's git wrappers.
from .task_git import _review_repo_context_sync  # noqa: F401

MAX_TASKS_PER_PLAN = 20

# Per-role model tiering. Running every cognition call on the account default
# (Opus) burns the Pro/Max quota fast and is slow; tier each role to the lightest
# model that does its job. These are `claude --model` values (an alias like
# 'sonnet'/'opus', or a full id). Planning is rare + high-leverage → Opus; the
# scope grill is conversational → Sonnet (set in elicitation.py); the eval judge
# is bounded classification → Haiku (set in eval_judge.py). The heavy coding path
# (OpenHands) is tiered separately via DEVCLAW_EXEC_MODEL. An empty value →
# the CLI's account default (no --model flag passed).
from .model_tiers import model_for as _model_for
PLANNER_MODEL = _model_for("planner")

VALID_KINDS: tuple[TaskKind, ...] = ("implement_feature", "fix_bug", "review_repository")


@dataclass
class PlannedTask:
    #: stable model-assigned id used to express deps within this plan only
    key: str
    goal: str
    kind: TaskKind
    #: keys (not UUIDs) of other tasks in this plan that must finish first
    depends_on_keys: list[str] = field(default_factory=list)
    #: the spec milestone this task serves (plan-from-spec only; else None)
    milestone: str | None = None


async def _plan_repo_context(workspace_dir: str) -> str:
    """Async wrapper — runs the blocking snapshot in a thread so it never blocks
    the event loop (same thread-offload rationale as task_queue's git wrappers).
    Looks up :func:`_review_repo_context_sync` as a module global so tests can
    patch it here."""
    return await asyncio.to_thread(_review_repo_context_sync, workspace_dir)


def build_planner_prompt(
    goal: str, workspace_dir: str, repo_context: str | None = None
) -> str:
    from .prompts import load_prompt

    parts = [load_prompt("plan-goal")]
    if repo_context and repo_context.strip():
        parts.append(
            "REPOSITORY CONTEXT (facts from the task workspace — the source of "
            "truth for repo identity and which files/dirs exist):\n"
            + repo_context.strip()
        )
    parts.append(
        f"Workspace: {workspace_dir}\n"
        f"Goal: {goal}\n\n"
        "Return the JSON now."
    )
    return "\n\n".join(parts)


def validate_plan(parsed: object) -> list[PlannedTask]:
    """Validate the parsed plan and return tasks in topological order. Raises
    PlannerError on cycles, dangling refs, missing fields, etc."""
    if not isinstance(parsed, dict):
        raise PlannerError("Plan must be a JSON object")
    raw = parsed.get("tasks")
    if not isinstance(raw, list):
        raise PlannerError("Plan.tasks must be an array")
    if len(raw) == 0:
        raise PlannerError("Plan must contain at least one task")
    if len(raw) > MAX_TASKS_PER_PLAN:
        raise PlannerError(
            f"Plan has {len(raw)} tasks; max is {MAX_TASKS_PER_PLAN}. Refine the goal."
        )

    seen: set[str] = set()
    tasks: list[PlannedTask] = []
    for t in raw:
        if not isinstance(t, dict):
            raise PlannerError("Each task must be an object")
        key = t.get("key").strip() if isinstance(t.get("key"), str) else ""
        goal = t.get("goal").strip() if isinstance(t.get("goal"), str) else ""
        kind_raw = t.get("kind") if isinstance(t.get("kind"), str) else "implement_feature"
        deps_raw = t.get("depends_on")
        milestone = t.get("milestone").strip() if isinstance(t.get("milestone"), str) else None
        if not key:
            raise PlannerError("Task missing 'key'")
        if not goal:
            raise PlannerError(f"Task '{key}' missing 'goal'")
        if key in seen:
            raise PlannerError(f"Duplicate task key '{key}'")
        if kind_raw not in VALID_KINDS:
            raise PlannerError(
                f"Task '{key}' has invalid kind '{kind_raw}'; "
                f"expected one of {', '.join(VALID_KINDS)}"
            )
        depends_on_keys: list[str] = []
        if deps_raw is not None:
            if not isinstance(deps_raw, list):
                raise PlannerError(f"Task '{key}' depends_on must be an array")
            for d in deps_raw:
                if not isinstance(d, str) or not d.strip():
                    raise PlannerError(f"Task '{key}' has non-string dep")
                if d == key:
                    raise PlannerError(f"Task '{key}' depends on itself")
                depends_on_keys.append(d.strip())
        seen.add(key)
        tasks.append(
            PlannedTask(
                key=key,
                goal=goal,
                kind=kind_raw,
                depends_on_keys=depends_on_keys,
                milestone=milestone or None,
            )
        )

    # Validate all dep refs resolve.
    for t in tasks:
        for d in t.depends_on_keys:
            if d not in seen:
                raise PlannerError(f"Task '{t.key}' depends on unknown key '{d}'")

    # Kahn topological sort — also detects cycles.
    by_key = {t.key: t for t in tasks}
    indegree = {t.key: len(t.depends_on_keys) for t in tasks}
    dependents: dict[str, list[str]] = {}
    for t in tasks:
        for d in t.depends_on_keys:
            dependents.setdefault(d, []).append(t.key)

    ready = sorted(k for k, n in indegree.items() if n == 0)
    ordered: list[PlannedTask] = []
    while ready:
        k = ready.pop(0)
        ordered.append(by_key[k])
        for d in dependents.get(k, []):
            indegree[d] -= 1
            if indegree[d] == 0:
                ready.append(d)
        ready.sort()  # deterministic order across runs

    if len(ordered) != len(tasks):
        raise PlannerError("Plan contains a dependency cycle")
    return ordered


#: planning (plan_goal + plan_spec) runs at the planner tier
_planner_caller = claude_with_model(PLANNER_MODEL, role="planner")


def _parse_plan(raw: str) -> list[PlannedTask]:
    """Extract → parse → validate a planner response into an ordered DAG."""
    json_text = extract_json(raw)
    try:
        parsed = json.loads(json_text)
    except json.JSONDecodeError as err:
        raise PlannerError(f"Planner JSON parse failed: {err}", raw) from err
    return validate_plan(parsed)


async def plan_goal(
    goal: str,
    workspace_dir: str,
    claude_caller: Callable[[str], Awaitable[str]] = _planner_caller,
) -> list[PlannedTask]:
    """Plan a bare goal string (the small-bounded `start_program` case).

    The prompt is grounded in a snapshot of the ACTUAL workspace (remote,
    key-file presence, tracked layout — :func:`~devclaw.task_git._review_repo_context_sync`,
    the #227 pattern): without it the prompt was byte-invariant between a
    populated repo and an empty scaffold target, and host-side ``claude``
    inherits devclaw's own checkout as cwd — the wrong-codebase contamination
    channel #227 closed for the review gate. Strictly best-effort: a snapshot
    hiccup degrades to an ungrounded prompt, it never fails planning."""
    try:
        repo_context: str | None = await _plan_repo_context(workspace_dir)
    except Exception as exc:  # noqa: BLE001 — best-effort, never fail planning
        sys.stderr.write(
            f"devclaw: planner repo snapshot failed (planning ungrounded): {exc}\n"
        )
        repo_context = None
    raw = await claude_caller(build_planner_prompt(goal, workspace_dir, repo_context))
    return _parse_plan(raw)


# ===== plan-from-spec ========================================================
# The build-a-project-from-scratch path: decompose an *approved spec* (the
# shared scope contract handed in by the OpenClaw waiter after scope_grill) into
# a milestone-ordered DAG. Richer than plan_goal — the model is grounded in the
# spec's milestones, acceptance criteria, scope, and constraints.

def build_spec_planner_prompt(spec: str, workspace_dir: str) -> str:
    from .prompts import load_prompt

    return (
        f"{load_prompt('plan-spec')}\n\n"
        f"Workspace: {workspace_dir}\n\n"
        f"APPROVED SPEC:\n{spec}\n\n"
        "Return the JSON now."
    )


async def plan_spec(
    spec: str,
    workspace_dir: str,
    claude_caller: Callable[[str], Awaitable[str]] = _planner_caller,
) -> list[PlannedTask]:
    """Decompose an approved spec into a milestone-ordered DAG. Same validated
    DAG shape as plan_goal, with per-task milestones populated."""
    raw = await claude_caller(build_spec_planner_prompt(spec, workspace_dir))
    return _parse_plan(raw)
